text/2020-02-04-auth.md

```diff
@@ -0,0 +1,206 @@
+* All TiKV clients must first authenticate with PD.
+* PD sends back a cryptographically signed token to the client. This means that the token cannot be tampered
+* This token may include as metadata the key range of allowed key access and any access restrictions (read-only)
+* The client then sends this token with it’s requests to TiKV
+* TiKV decrypts the token.
+  * Failure: A decryption failure means the token is expired or was tampered with, and the client is denied access.
+  * Success: The client is authenticated.
+    * TiKV performs authorization by looking at the included metadata to ensure that the client is operating on the keys that it is allowed to. If not, the request is denied.
+
+
+#### Delegation for tooling
+
+Database interfaces like TiDB that can implement their own auth.
+But now we have tools that should be using that auth but need direct TiKV access.
+
+This can be implemented via delegation. A tool would:
+* login to TiDB
+* request a signed token for accessing TiKV
+* TiDB gives a token to the tool that will be recognized by TiKV as delegated by the TiDB user
+* The token also includes access restrictions given by TiDB: allowed key range and permissions such as read-only
+* This token can then be sent to TIKV just as TiDB itself sends a token
+
+Tools such as backup could benefit from a read-only access restriction.
+
+This proposal does not seek to attempt to push down TiDB auth into the TiKV layer.
+Database layers should implement their own auth.
+This could cause some difficulties for mapping database access restrictions to TiKV restrictions, but in theory if they know how to store data then they should know how to restrict access.
+
+
+
+#### Reducing overhead
+
+Auth will create some additional latency for new requests.
+However, I don’t think it needs to create any noticeable overhead for long-lived connections.
+TiDB gets a token from PD that can be used for some time period (perhaps 15 minutes).
+In the background during that 15 minutes TiDB will renew its token and then switch to the renewed token.
+If a long-lived connection can switch to the new token while still connected, then we can avoid disrupting long-lived connections.
+Short-lived connections and re-connections can also be optimized In TiKV by caching successfully decrypted tokens.
+
+
+#### Shared secret and bootstraping
+
+In the above scenario, TiKV must have a mechanism to be able to decrypt the token and ensure it was not tampered with.
+One way of doing this is to encrypt it with a secret that only TiKV knows.
+At startup, TiKV can get the secret from PD.
+However, we must be able to prevent impersonation of a TiKV node, and we can do this with our existing auth mechanism
+
+TiKV itself must also authenticate.
```

Comment:
This could be done by the `put_store` operation from TiKV to PD. The token could be refreshed by the following heartbeat.